### PR TITLE
Allow detection of taps on TabBar

### LINKED ISSUE #24224: TabBar does not allow detection of taps
# Problem

We have come up against a wall in our project because of this a couple times now. The latest example is as follows:

We have a `TabBar `with two `Tab`s, both of which contain lists. If we are in Tab X, and the user taps on the header for Tab X, we want to scroll the list up to the top. If the user taps Tab Y, we of course want to animate there. This currently seems to be impossible. We can wrap a `GestureDetector` around a `Tab` child of `TabBar` and sort of intercept taps, but this only works for the actual icon and text of the Tab, not the padding around it. Moreover, it disables the normal reaction to taps on the icon and text (tabs no longer switch when tapping there).

Another example would be bookkeeping of the current tab index. The `TabController` does allow for access to the index, and we can be notified by subscribing to the `TabController`, but it cannot be trusted because of the following behaviour:

When tapping on a tab:

1. Notified
2. Widget is built
3. Notified

When swiping to a tab:

1. Widget is built
2. Notified

This means that it is impossible to react to swipes and we always need to be aware of the fact that we might currently be in a swipe, but the system is simply not telling us. As a result, we cannot keep track of
the current index of tabs, since it can be corrupted by a simple swipe (and any workarounds are unnecessarily complicated).

Finally, while this is not something we need, it is also currrently impossible to disable tapping on the `TabBar`, see this [SO question](https://stackoverflow.com/questions/51905079/disable-ontapped-on-tabbar-on-flutter).

# Fix
See PR #23919

### PR DESCRIPTION
**Pull request by JaspervanRiet**
Fixes #24224 

I have added a callback to the `TabBar` that is called if not null. This would fix all 3 of these examples.

### CODE REVIEW

**Comment by Hixie** [comment]
This will need a test.

**Comment by Hixie** [comment]
Generally speaking I'd like to keep callbacks such that not setting one and setting one to a no-op callback (as in `() { }`) do the same thing (i.e. adding a no-op callback shouldn't change behaviour). It can be a bit unintuitive if adding something that does nothing does something. :-)

Not really sure what the best way to do that here would be, though...

cc @HansMuller

**Comment by JaspervanRiet** [comment]
I have opened an issue so there can be proper discussion (should have done this before) #24224. A potential solution would be to handle this in the `_handleTap` function, first let the TabBar handle the tap, and then invoke a callback to the user if non null.

File: packages/flutter/lib/src/material/tabs.dart [hunk 1]
```diff
@@ -660,6 +661,9 @@ class TabBar extends StatefulWidget implements PreferredSizeWidget {
   /// is null then the text style of the theme's body2 definition is used.
   final TextStyle unselectedLabelStyle;
 
+  /// An option callback when the [TabBar] is tapped.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/tabs.dart]: An optional callback that's called when the [TabBar] is tapped.

The callback is applied to the index of the tab where the tap occurred. 

This callback has no effect on the default handling of taps. Applications that want to do a little extra work when a tab is tapped, even if the tap doesn't change the TabController's index, can specify [onTap].

File: packages/flutter/test/material/tabs_test.dart [hunk 2]
```diff
@@ -1758,6 +1758,68 @@ void main() {
...
+      String value,
+      bool isScrollable = false,
+      Color indicatorColor,
+    }) {
+      return boilerplate(
+        child: Container(
+          child: TabBar(
+            controller: controller,
+            tabs: tabs.map<Widget>((String tab) => Tab(text: tab)).toList(),
+              onTap: (int index) {
```
Comment (HansMuller) [inline packages/flutter/test/material/tabs_test.dart]: need to fix the indent here

File: packages/flutter/test/material/tabs_test.dart [hunk 3]
```diff
@@ -1758,6 +1758,68 @@ void main() {
...
+                if (controller.index != index) {
```
Comment (HansMuller) [inline packages/flutter/test/material/tabs_test.dart]: This is what we don't want developers to do; onTap is a "me too" notification. 

Maybe this is leftover from a previous version of the PR?

File: packages/flutter/test/material/tabs_test.dart [hunk 4]
```diff
@@ -1758,6 +1758,68 @@ void main() {
     semantics.dispose();
   });
 
+  testWidgets('can be notified of TabBar onTap behavior', (WidgetTester tester) async {
```
Comment (HansMuller) [inline packages/flutter/test/material/tabs_test.dart]: This test looks like it needs to be revised. It should confirm that the onTap() call is applied to the correct index and that it happens even when the default tap handler doesn't have any effect on the TabController.

**Comment by JaspervanRiet** [comment]
@HansMuller Thanks for the review, I believe these have now been resolved.

File: packages/flutter/lib/src/material/tabs.dart [hunk 5]
```diff
@@ -661,7 +661,15 @@ class TabBar extends StatefulWidget implements PreferredSizeWidget {
...
   final TextStyle unselectedLabelStyle;
 
-  /// An option callback when the [TabBar] is tapped.
+  /// An optional callback that's called when the [TabBar] is tapped.
+  ///
+  /// The callback is applied to the index of the tab where the tap occurred.
+  ///
+  /// This callback has no effect on the default handling of taps. It's for
+  /// applications that want to do a little extra work when a tab is tapped,
+  ///  even if the tap doesn't change the TabController's index. TabBar [onTap]
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/tabs.dart]: extra leading space on this line and the next

File: packages/flutter/test/material/tabs_test.dart [hunk 6]
```diff
@@ -1788,36 +1787,66 @@ void main() {
...
+    expect(tabIndex, controller.index);
+
+    hasBeenTapped = false;
+    tabIndex = -1;
 
-    await tester.pumpWidget(buildFrame(tabs: tabs, value: 'C', isScrollable: false, controller: controller));
+    await tester.pumpWidget(buildFrame(tabs: tabs, controller: controller));
     await tester.tap(find.text('C'));
     await tester.pump();
     await tester.pump(const Duration(seconds: 1));
```
Comment (HansMuller) [inline packages/flutter/test/material/tabs_test.dart]: tester.pumpAndSetttle() here and below

File: packages/flutter/test/material/tabs_test.dart [hunk 7]
```diff
@@ -1758,6 +1758,97 @@ void main() {
     semantics.dispose();
   });
 
+  testWidgets('can be notified of TabBar onTap behavior', (WidgetTester tester) async {
+    int tabIndex = -1;
+    bool hasBeenTapped = false;
```
Comment (HansMuller) [inline packages/flutter/test/material/tabs_test.dart]: This appears to be redundant with tabIndex != -1 and existing `expect`s already check that.

**Review by HansMuller — COMMENTED** [review]
This looks good, just needs a few trivial changes.

**Review by HansMuller — APPROVED** [review]
LGTM

I will land this as soon as the repo is open again.

**Pull request merged**